openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,323 @@
+		return admission.NewForbidden(attr, err) // this should not happen though
+	}
+
+	if !isNamespaceAllowed(ns.Annotations) {
+		// strip any annotations that can affect underlying components to apply cpuset and cpushares
+		stripResourcesAnnotations(pod.Annotations)
+		stripWorkloadAnnotations(pod.Annotations)
+
+		return nil
+	}
+
+	if !isPodManaged(pod.Annotations) {
+		// strip any annotations that can affect underlying components to apply cpuset and cpushares
+		stripResourcesAnnotations(pod.Annotations)
+
+		return nil
+	}
+
+	nodes, err := a.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.Warningf("%s got an error retrieving nodes: %v", PluginName, err)
+		return admission.NewForbidden(attr, err) // this should not happen though
+	}
+
+	if !isManagementResourceAvailable(nodes) {
+		klog.Warningf("%s no nodes with resource %s is available", PluginName, managedCapacityLabel)
+
+		// strip any annotations that can affect underlying components to apply cpuset and cpushares
+		stripResourcesAnnotations(pod.Annotations)
+		stripWorkloadAnnotations(pod.Annotations)
+		return nil
+	}
+
+	allContainers := append([]coreapi.Container{}, pod.Spec.InitContainers...)
+	allContainers = append(allContainers, pod.Spec.Containers...)
+	podQoSClass := getPodQoSClass(allContainers)
+
+	// we do not want to change guaranteed pods resource allocation, because it should be managed by
+	// relevant managers(CPU and memory) under the kubelet
+	if podQoSClass == coreapi.PodQOSGuaranteed {
+		klog.Warningf("Skip pod %q CPUs requests modifications under namespace %q because it has guaranteed QoS class", pod.Name, pod.Namespace)
+		return nil
+	}
+
+	// before we update the pod available under admission attributes, we need to verify that deletion of the CPU request
+	// will not change the pod QoS class
+	// 1. Copy the pod
+	// 2. Delete CPUs requests for all containers under the pod
+	// 3. Get modified pod QoS class
+	// 4. Verify that the pod QoS class before and after the modification stay the same
+	// 5. Update the pod under admission attributes
+	podCopy := pod.DeepCopy()
+	updatePodResources(podCopy, podQoSClass)
+
+	allContainersCopy := append([]coreapi.Container{}, podCopy.Spec.InitContainers...)
+	allContainersCopy = append(allContainersCopy, podCopy.Spec.Containers...)
+	podQoSClassAfterModification := getPodQoSClass(allContainersCopy)
+
+	if podQoSClass != podQoSClassAfterModification {
+		klog.Warningf("Skip pod %q CPUs requests modifications because it will change the pod QoS class from %s to %s", pod.Name, podQoSClass, podQoSClassAfterModification)
+		return nil
+	}
+
+	updatePodResources(pod, podQoSClass)
+
+	return nil
+}
+
+func updatePodResources(pod *coreapi.Pod, class coreapi.PodQOSClass) {
+	if pod.Annotations == nil {
+		pod.Annotations = map[string]string{}
+	}
+
+	// remove all relevant resource annotations, to prevent from a user to apply unwanted changes
```

Comment:
what does this mean? A user can readd the annotations easily. How does stripping help?

Reply:
the intention is to remove resource and workload annotations that a user can add manually